### PR TITLE
Topics with underscore have invalid metrics paths in zookeeper

### LINKED ISSUE #740: Invalid metrics names in zookeeper for topics with underscore in name
When topic has underscore char in name then its counters are kept in zookeeper under wrong path (underscores are escaped to dots). Hence, for this topic:

- counter values are not read properly so user sees 0 for published, delivered, discarded counters
- counted number of consumers assigned to subscription is always 1. Because of this calculated max-rate is too high

### PR DESCRIPTION
**Pull request by druminski**
Resolves #740

### CODE REVIEW

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ private static TopicName escapedTopicName(String qualifiedTopicName) {
         TopicName topicName = fromQualifiedName(qualifiedTopicName);
         return new TopicName(
                 escapeMetricsReplacementChar(topicName.getGroupName()),
-                escapeMetricsReplacementChar(topicName.getName())
+                topicName.getName()
```
Comment (adamdubiel) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java]: will it be backwards compatible? what with all the metrics we already collected?

**Comment by chemicL** [comment]
If I remember correctly, subscription name could also be a problem.
Writing:
    escapeMetricsReplacementChar(matcher.getSubscriptionName()),
Reading:
    subscription.getName()

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java (hunk 1 above)
Comment (druminski) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java]: Yes, this change is backward compatible. It only touches broken metrics which actually were not read by Hermes (symptoms described here #740). Topics without underscore in names have valid metrics names and nothing will change for them.

Comment (adamdubiel) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java]: okay, one more: why do we fix topic name and not group name? what if group name has an underscore?

**Comment by druminski** [comment]
@chemicL no, it is not a problem for subscription because hermes-console does not allow to use underscore char in subscription name. I also checked our production state and confirm that we don't have subscriptions with underscores.

As for
```
escapeMetricsReplacementChar(matcher.getSubscriptionName())
```
we have to do this because ZookeeperCounterReporter reads metrics from MetricRegistry and subs there are escaped from `.` to `_`. So to keep original sub name we have to escape it again before send it to zookeeper.

File: hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java (hunk 1 above)
Comment (druminski) [inline hermes-common/src/main/java/pl/allegro/tech/hermes/common/metric/counter/zookeeper/ZookeeperCounterReporter.java]: Currently we don't have underscores in groups names on prod and in this PR I am adding restriction to proper field in hermes-console, so this situation will not occur in the future. 

When metric is created in MetricRegistry then Hermes escapes group, topic and subscription names from `.` to `_`. However topic name can't have dots so this transformation will do nothing for topic name. This is the reason why we can use underscore in topic name but not in group and subscription name. In ZookeeperCounterReporter Hermes again transforms `_` to `.` to keep original names but it shouldn't do this for topic name.

**Comment by chemicL** [comment]
Great, I also checked if we allow underscores in subscription names: we do not. Let's go :-)

**Comment by adamdubiel** [comment]
Will it fix #448  as well?

**Comment by druminski** [comment]
@adamdubiel yes, after this fix structures for metrics in zk will be created correctly. However, to keep neatness in zookeeper we have to remove those broken one.

**Comment by adamdubiel** [comment]
Ok, but i can close #448 as related problem.

**Comment by druminski** [comment]
sure

**Pull request merged**